### PR TITLE
Add satellite basemap

### LINKED ISSUE #36: Add ESRI Satellite Basemap Layer
## Overview

Currently there is only one basemap, the gray ESRI layer. We need to add the satellite layer as well.

- [x] Select which ESRI Staellite basemap to use
- [x] Add it to the map
- [x] Wire up the basemap selector so that it switches between the layers correctly

### PR DESCRIPTION
**Pull request by mstone121**
## Overview

This PR adds the satellite basemap and implements dynamic switching with the basemap setting in state.

Closes #36 

### Notes

`react-leaflet` does not allow the `url` prop on TileLayers to be mutable. This PR implements custom basemap components that add/remove layers in sync with the React element lifecycle.

## Testing Instructions

- On welcome page:
  - [x] Ensure background basemap is unaffected
- On draw map page:
  - [x] Switch between the two basemap types from the sidebar

## Checklist

- [x] `fixup!` commits have been squashed
- [x] `CHANGELOG.md` updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [x] `README.md` updated if necessary to reflect the changes
- [x] CI passes after rebase

### CODE REVIEW

**Comment by rajadain** [comment]
Taking a look now

**Comment by rajadain** [comment]
This works really well, nice job figuring out how to do it in a react-y way.

Since we have multiple pieces of information related to a given basemap, I think we should try to put them in a central place. Something like this perhaps:

```diff
diff --git a/src/app/src/components/Basemaps.js b/src/app/src/components/Basemaps.js
index 6c37fbc..eabcdbe 100644
--- a/src/app/src/components/Basemaps.js
+++ b/src/app/src/components/Basemaps.js
@@ -32,3 +26,21 @@ function Basemap({ url, attribution }) {
         };
     }, [map, url, attribution]);
 }
+
+export const BASEMAPS = {
+    default: {
+        component: (
+            <Basemap url={DEFAULT_BASEMAP_URL} attribution={ESRI_ATTRIBUTION} />
+        ),
+        maxZoom: 16,
+    },
+    satellite: {
+        component: (
+            <Basemap
+                url={SATELLITE_BASEMAP_URL}
+                attribution={ESRI_ATTRIBUTION}
+            />
+        ),
+        maxZoom: 18,
+    },
+};
```

```diff
diff --git a/src/app/src/components/DrawMap/DrawMap.js b/src/app/src/components/DrawMap/DrawMap.js
index 7a21e48..7c6609c 100644
--- a/src/app/src/components/DrawMap/DrawMap.js
+++ b/src/app/src/components/DrawMap/DrawMap.js
@@ -2,7 +2,7 @@ import { Button, Icon } from '@chakra-ui/react';
 import { useSelector } from 'react-redux';
 import { ArrowRightIcon } from '@heroicons/react/outline';
 
-import { DefaultBasemap, SatelliteBasemap } from '../Basemaps';
+import { BASEMAPS } from '../Basemaps';
 
 import Map from '../Map';
 import EditToolbar from './EditToolbar';
@@ -25,14 +25,11 @@ export default function DrawMap() {
 function Basemap() {
     const basemapType = useSelector(state => state.map.basemapType);
 
-    switch (basemapType) {
-        case 'default':
-            return <DefaultBasemap />;
-        case 'satellite':
-            return <SatelliteBasemap />;
-        default:
-            throw Error(`Invalid basemap type: $basemapType`);
+    if (!(basemapType in BASEMAPS)) {
+        throw Error(`Invalid basemap type: ${basemapType}`);
     }
+
+    return BASEMAPS[basemapType].component;
 }
 
 function SaveAndBackButton() {
diff --git a/src/app/src/components/DrawMap/MapFeatures.js b/src/app/src/components/DrawMap/MapFeatures.js
index fe825d6..794ca1d 100644
--- a/src/app/src/components/DrawMap/MapFeatures.js
+++ b/src/app/src/components/DrawMap/MapFeatures.js
@@ -2,6 +2,7 @@ import { useEffect } from 'react';
 import { useSelector } from 'react-redux';
 import { useMap } from 'react-leaflet';
 
+import { BASEMAPS } from '../Basemaps';
 import useEditingPolygon from './useEditingPolygon';
 
 export default function MapFeatures() {
@@ -16,12 +17,8 @@ function useSetMaxZoomLevel() {
     const map = useMap();
 
     useEffect(() => {
-        switch (basemapType) {
-            case 'default':
-                map.setMaxZoom(16);
-                break;
-            default:
-                map.setMaxZoom(18);
+        if (basemapType in BASEMAPS) {
+            map.setMaxZoom(BASEMAPS[basemapType].maxZoom);
         }
     }, [basemapType, map]);
 }
```

This looks odd though:

```diff
diff --git a/src/app/src/pages/Welcome.js b/src/app/src/pages/Welcome.js
index 73563f0..9b5e48d 100644
--- a/src/app/src/pages/Welcome.js
+++ b/src/app/src/pages/Welcome.js
@@ -1,14 +1,12 @@
 import Map from '../components/Map';
 import WelcomeModal from '../components/WelcomeModal';
-import { DefaultBasemap } from '../components/Basemaps';
+import { BASEMAPS } from '../components/Basemaps';
 
 export default function Welcome() {
     return (
         <>
             <WelcomeModal />
-            <Map>
-                <DefaultBasemap />
-            </Map>
+            <Map>{BASEMAPS['default'].component}</Map>
         </>
     );
 }
```

so perhaps this approach can't be taken all the way, and there is still value in having `DefaultBasemap` be a component.

Any thoughts on this arrangement?

**Comment by mstone121** [comment]
> Any thoughts on this arrangement?

It's not an approach I'm familiar with, but it seems like it would work. When the `BASEMAPS` object is imported, it doesn't render those components, right? I would assume just the function reference is imported and it won't be run until the virtual DOM finds it. 

Alternatively, we could incorporate the zoom level into the `<Basemap />` component as a prop:
```jsx
export function DefaultBasemap() {
    return (
        <Basemap
            url={DEFAULT_BASEMAP_URL}
            attribution={ESRI_ATTRIBUTION}
            maxZoomLevel={16}
        />
    );
}

export function SatelliteBasemap() {
    return (
        <Basemap
            url={SATELLITE_BASEMAP_URL}
            attribution={ESRI_ATTRIBUTION}
            maxZoomLevel={18}
        />
    );
}

function Basemap({ url, attribution, maxZoomLevel }) {
    useSetMaxZoomLevel(maxZoomLevel);
    const map = useMap();

    useEffect(() => {
        const layer = L.tileLayer(url, { attribution });
        map.addLayer(layer);

        return () => {
            if (map.hasLayer(layer)) {
                map.removeLayer(layer);
            }
        };
    }, [map, url, attribution]);
}

```
Adding the zoom level parameter to `useMaxZoomLevel` should be simple.

**Comment by rajadain** [comment]
I like the approach you outlined above. Let's try that.

**Comment by mstone121** [comment]
> I like the approach you outlined above. Let's try that.

[aa0337f](https://github.com/azavea/iow-boundary-tool/pull/39/commits/aa0337f3c120d9df210407ea7cbc79946e6e2964)

**Comment by rajadain** [comment]
The code looks good. What is the source of the new green "Default" map thumbnail? It doesn't look like the gray basemap.

**Review by rajadain — COMMENTED** [review]
The code looks good. What is the source of the new green "Default" map thumbnail? It doesn't look like the gray basemap.

File: src/app/src/components/DrawMap/DrawMap.js [hunk 1]
```diff
@@ -19,6 +22,19 @@ export default function DrawMap() {
...
+function Basemap() {
+    const basemapType = useSelector(state => state.map.basemapType);
+
+    switch (basemapType) {
+        case 'default':
+            return <DefaultBasemap />;
+        case 'satellite':
+            return <SatelliteBasemap />;
+        default:
+            throw Error(`Invalid basemap type: $basemapType`);
```
[the remark this answers is not in the document]
Comment (mstone121) [reply src/app/src/components/DrawMap/DrawMap.js]: https://github.com/azavea/iow-boundary-tool/pull/39/commits/cb6488902db88e8faedcd3cf6da64d43f9f4dee8

**Comment by mstone121** [comment]
> What is the source of the new green "Default" map thumbnail? It doesn't look like the gray basemap.

It's what Gates sent in the slack channel. I think it's closer to what the ideal basemap would look like.

**Review by rajadain — APPROVED** [review]
+1 tested, nice clean implementation!

**Pull request merged**